homeassistant/components/number/__init__.py

```diff
@@ -94,6 +97,11 @@ def step(self) -> float:
                 step /= 10.0
         return step
 
+    @property
+    def mode(self) -> str:
```

Comment:
Isn't this competing with device class?

Reply:
How is that?

Reply:
It's an attribute that describes the properties of this entity which influences how it is represented.

Reply:
That is true, so is the device class, icon, name, picture and such itself.
There must be a way for an integration to provide details on e.g., min/max, but also a representation. I could see the frontend being able to override that (e.g., in a Lovelace configuration).

Alternatively, we could indeed introduce device classes to the `number` integration, although, in general, the input is not always related to the unit shown.

E.g., Speed. Could be something between 0-100%/1-255, or something between 500-15000rpm. The first would be a nice slider, the second... probably not that much.

Reply:
I think that we can also start with a nice algorithm. It would also benefit the selectors. Maybe something like:

```
const possibleValues = (max - min) / steps;
const useSlider = steps <= 255;
```

Reply:
Correct me if I'm mistaken, but I assume that these changes are beyond the scope of this pull request. 

Are there any changes necessary to get this merged in?

Reply:
It would be an alternative that would render this PR not needed.

Reply:
The advantage of the mode attribute is that it gives a platform more control over the representation, and the platform knows better whether the value it has is more discrete or monotonous. Also, it gives the control for the user to override it using the existing `customize` functionality.

As a compromise for the both, I can suggest to implement the same nice algorithm but as a default value for the attribute, e.g. instead of `return MODE_SLIDER` it would be more like `return MODE_SLIDER if (max - min) / steps <= 255 else MODE_BOX`.

Reply:
> Also, it gives the control for the user to override it using the existing customize functionality.

That is not something that should be used to adjust UI behavior, for these things we have the UI itself.

I agree with the suggestion given by balloob for that matter.

Reply:
We should start with my suggestion and then we can see if this PR is still needed. Let's not apply multiple solutions at once.